text/0000-dyn-error-generic-member-access.md

```diff
@@ -0,0 +1,471 @@
+    /// and `Ok(self)` if `T` was not the type being requested.
+    ///
+    /// The passed closure is only called if the value will be successfully
+    /// provided.
+    ///
+    /// This method can be chained within `provide` implementations using the
+    /// `?` operator to concisely provide multiple objects.
+    pub fn provide_with<T: ?Sized + 'static, F>(mut self, cb: F) -> ProvideResult<'r, 'out>
+    where
+        F: FnOnce() -> &'out T,
+    {
+        match self.downcast_buf::<T>() {
+            Some(this) => {
+                debug_assert!(
+                    this.value.is_none(),
+                    "Multiple requests to a `RequestBuf` were acquired?"
+                );
+                this.value = Some(cb());
+                Err(FulfilledRequest(PhantomData))
+            }
+            None => Ok(self),
+        }
+    }
+
+    /// Get the `TypeId` of the requested type.
+    pub fn type_id(&self) -> TypeId {
+        unsafe { *self.buf.as_ref() }
+    }
+
+    /// Returns `true` if the requested type is the same as `T`
+    pub fn is<T: ?Sized + 'static>(&self) -> bool {
+        self.type_id() == TypeId::of::<T>()
+    }
+
+    /// Try to downcast this `Request` into a reference to the typed
+    /// `RequestBuf` object.
+    ///
+    /// This method will return `None` if `self` was not derived from a
+    /// `RequestBuf<'_, T>`.
+    fn downcast_buf<T: ?Sized + 'static>(&mut self) -> Option<&mut RequestBuf<'out, T>> {
+        if self.is::<T>() {
+            unsafe { Some(&mut *(self.buf.as_ptr() as *mut RequestBuf<'out, T>)) }
+        } else {
+            None
+        }
+    }
+
+    /// Calls the provided closure with a request for the the type `T`, returning
+    /// `Some(&T)` if the request was fulfilled, and `None` otherwise.
+    ///
+    /// The `ObjectProviderExt` trait provides helper methods specifically for
+    /// types implementing `ObjectProvider`.
+    pub fn with<T: ?Sized + 'static, F>(f: F) -> Option<&'out T>
+    where
+        for<'a> F: FnOnce(Request<'a, 'out>) -> ProvideResult<'a, 'out>,
+    {
+        let mut buf = RequestBuf {
+            type_id: TypeId::of::<T>(),
+            value: None,
+        };
+        let _ = f(Request {
+            buf: unsafe {
+                NonNull::new_unchecked(&mut buf as *mut RequestBuf<'out, T> as *mut TypeId)
+            },
+            _marker: PhantomData,
+        });
+        buf.value
+    }
+}
+
+// Needs to have a known layout so we can do unsafe pointer shenanigans.
+#[repr(C)]
+struct RequestBuf<'a, T: ?Sized> {
+    type_id: TypeId,
+    value: Option<&'a T>,
+}
+
+/// Marker type indicating a request has been fulfilled.
+pub struct FulfilledRequest(PhantomData<&'static Cell<()>>);
+
+/// Provider method return type.
+///
+/// Either `Ok(Request)` for an unfulfilled request, or `Err(FulfilledRequest)`
+/// if the request was fulfilled.
+pub type ProvideResult<'r, 'a> = Result<Request<'r, 'a>, FulfilledRequest>;
```

Comment:
This looks really odd that `Err` is the "happy case" especially as a public type. If the purpose is just to use `?` chaining it is probably better to just create an `impl Try` type like [LoopState](https://github.com/rust-lang/rust/blob/2454a68cfbb63aa7b8e09fe05114d5f98b2f9740/src/libcore/iter/mod.rs#L375).

```rust
enum Answer<T, S> {
    Found(T),
    Continue(S),
}
```